tmt/utils.py

```diff
@@ -4787,11 +4796,18 @@ def git_clone(
                 url, destination
                 ), env=env)
     except RunError:
+        if retry_times == 0:
+            raise
         if not shallow:
             # Do not retry if shallow was not used
             raise
         # Git server might not support shallow cloning, try again (do not modify url)
-        return git_clone(url, destination, common, env, shallow=False, can_change=False)
+        shallow = retry_times > 1
```

Comment:
Why is `shallow` changed at this point?

Reply:
If `shallow` is always `False`, it will hit `L4790-L4791`, then we just have retried once.

```python
4789     except RunError:                                                                                
4790         if not shallow:                                                                             
4791             # Do not retry if shallow was not used                                                  
4792             raise  
```

Reply:
I see. But, IIUIC, this line needs to be smarter: if `git_clone()` is called with `shallow=False`, this check would switch `shallow` to `True` for the first retry, right?  `git_clone()` docs state that if `shallow` was requested, a retry would be run without it, but makes no statements about what happens when `shallow` was not requested - this change would enable `shallow` for retries except the very last one. I guess this needs to check whether `shallow is True`, and never touch `shallow` if it was disabled.

Reply:
> this change would enable shallow for retries except the very last one. I guess this needs to check whether shallow is True, and never touch shallow if it was disabled.

* If `shallow is False` at the first time, the first retry of  `git_clone()` will never happen because of L4790.
* If `shallow is True` at the first time, retry will happen if we use `shallow = retries > 0`. And when `retries == 0`, `shallow is False` and `retry` will be stopped in the last call.

Hi @happz, I'm a bit confused about what you say on `...shallow is True, and never touch shallow if it was disabled.` Would you please help to explain more?

Reply:
Sure. Imagine how `git_clone` can be called:

```python
# shallow == False
git_clone(..., shallow=False, ...)
```

There would be no retries in this case, because of the `if not shallow` test on line 4804. IIUIC, there should be retries even for this case

```python
# shallow == True
git_clone(..., shallow=True, ...)
```

In this case, first time retrying, `git_clone()` would call `git_clone()` with `shallow = retries > 0` - but this will **always** be `True`, right? The conditional setting of `shallow` is not needed because, for any value bigger than zero, this will always yield `True`, and the zero value would never get here because of `if not retries` test on line 4802. So, the assignment can be dropped, and we can use `True` directly:

```python
        return git_clone(url, destination, common, env,
                         shallow=True,
                         can_change=False,
                         ...
```

Reply:
Thank you for your explanation. It's clear to me :-)  And the related code is updated, please help to review, thanks!

```patch
-        shallow = retries > 0
         return git_clone(url, destination, common, env,
-                         shallow=shallow,
+                         shallow=True,
                          can_change=False,
                          retries=retries - 1,
                          timeout=timeout)
```

Reply:
Hi @psss, with [the solution][01] suggested by you, `retry` is done actually.  For details, please take a look at the [debugging output][02].

[01]: https://github.com/idorax/tmt-dev-sandbox/blob/dev.20231024.2317/2317/01/utils.patch.md
[02]: https://github.com/idorax/tmt-dev-sandbox/blob/dev.20231024.2317/2317/01/debug.md